File: src/04.stack/13.394.go
package stack

import "strings"

func DecodeString(s string) string {
	// 辅助栈
	// res := ""
	// stackStr := []string{}
	// stackNum := []int{}
	// num := 0

	// for _, v := range s {
	// 	if v == '[' {
	// 		stackNum = append(stackNum, num)
	// 		stackStr = append(stackStr, res)
	// 		num = 0
	// 		res = ""
	// 	} else if v == ']' {
	// 		lastNum := stackNum[len(stackNum)-1]
	// 		lastStr := stackStr[len(stackStr)-1]
	// 		res = lastStr + strings.Repeat(res, lastNum)
	// 		stackNum = stackNum[:len(stackNum)-1]
	// 		stackStr = stackStr[:len(stackStr)-1]
	// 	} else if '0' <= v && v <= '9' {
	// 		num = num*10 + int(v-'0')
	// 	} else {
	// 		res += string(v)
	// 	}
	// }

	// return res

	_, val := searchHandle(s, 0)
	return val
}

func searchHandle(s string, i int) (index int, val string) {
	index = 0
	val = ""
	res := ""
	num := 0
	for i < len(s) {
		if s[i] == '[' {
			index, val = searchHandle(s, i+1)
			i = index
			res += strings.Repeat(val, num)
			num = 0
		} else if s[i] == ']' {
			val = res
			index = i
			return
		} else if '0' <= s[i] && s[i] <= '9' {
			num = num*10 + int(s[i]-'0')
		} else {
			res += string(s[i])
		}
		i++
	}
	val = res
	index = i
	return
}

File: src/04.stack/13.394.ts
export function decodeString(s: string): string {
    // 辅助栈:
    // const zero = "0".charCodeAt(0);
    // const nine = "9".charCodeAt(0);
    // const stack = [];
    // let curVal = 0;
    // let res = "";
    // for (let i = 0; i < s.length; i++) {
    //     if (s[i] === "[") {
    //         stack.push([curVal, res]);
    //         res = "";
    //         curVal = 0;
    //     } else if (s[i] === "]") {
    //         const item = stack.pop();
    //         res = item[1] + res.repeat(item[0]);
    //     } else if (zero <= s.charCodeAt(i) && s.charCodeAt(i) <= nine) {
    //         curVal = curVal * 10 + s.charCodeAt(i) - zero;
    //     } else {
    //         res += s[i];
    //     }
    // }
    // return res;

    const zero = "0".charCodeAt(0);
    const nine = "9".charCodeAt(0);
    const search = (i: number) => {
        let res = "";
        let curVal = 0;
        while (i < s.length) {
            if (s[i] === "[") {
                const { index, val } = search(i + 1);
                i = index;
                res += val.repeat(curVal);
                curVal = 0;
            } else if (s[i] === "]") {
                return {
                    val: res,
                    index: i,
                };
            } else if (zero <= s.charCodeAt(i) && s.charCodeAt(i) <= nine) {
                curVal = curVal * 10 + s.charCodeAt(i) - zero;
            } else {
                res += s[i];
            }
            i++;
        }

        return {
            val: res,
            index: i,
        };
    };

    return search(0).val;
}
